src/hotspot/share/utilities/exceptions.cpp

```diff
@@ -262,8 +263,24 @@ void Exceptions::fthrow(JavaThread* thread, const char* file, int line, Symbol*
   va_list ap;
   va_start(ap, format);
   char msg[max_msg_size];
-  os::vsnprintf(msg, max_msg_size, format, ap);
+  int ret = os::vsnprintf(msg, max_msg_size, format, ap);
   va_end(ap);
+  // If ret == -1 then either there was a format conversion error, or the required buffer size
+  // exceeds INT_MAX and so couldn't be returned (undocumented behaviour of vsnprintf). Depending
+  // on the platform the buffer may be filled to its capacity (Linux), filled to the conversion
+  // that encountered the overflow (macOS), or is empty (Windows), so it is possible we
+  // have a truncated UTF-8 sequence. Similarly, if the buffer was too small and ret >= max_msg_size
+  // we may also have a truncated UTF-8 sequence. In such cases we need to fix the buffer so the UTF-8
+  // sequence is valid.
+  if ((ret == -1 || ret >= max_msg_size) && strlen(msg) > 0) {
+    assert(msg[max_msg_size - 1] == '\0', "should be null terminated");
+    UTF8::truncate_to_legal_utf8((unsigned char*)msg, max_msg_size);
```

Comment:
Ah, I misread your patch and thought you pass in the strlen of the message to the truncation function, when in fact you pass in the hard coded message buffer size. 

But that begs the question of why you test strlen above, and more importantly, whether all cases where snprintf returns an error are truncation problems. It could have detected an invalid UTF8 sequence and aborted in the middle of it.

Reply:
The `strlen` check is to skip the empty buffer you can get on Windows if vsnprintf returns -1 due to overflow of INT_MAX.

We are assuming/requiring that we start with a valid UTF8 sequence and the worst that will happen is that vsnprintf will truncate it.

If we actually got -1 for a conversion error (no way to tell the difference in the two cases) then we would unnecessarily truncate, but we do not expect any such conversion errors - in part because we type check the format specifiers and args and so should never get a mismatch.

Reply:
Thanks for the off-list discussion @djelinski , I now understand what you mean here. Code updated and commented.

Reply:
Note this has been updated now to pass `strlen(msg)`.